### PR TITLE
Commit-Reveal Weights 

### PR DESCRIPTION
**Pull request by JohnReedV**
## Related Issue(s)
[Issue#389](https://github.com/opentensor/subtensor/issues/389)

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Other (please describe):

## Breaking Change

If this PR introduces a breaking change, please provide a detailed description of the impact and the migration path for existing applications.

## Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have run `cargo fmt` and `cargo clippy` to ensure my code is formatted and linted correctly
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

## Screenshots (if applicable)

Please include any relevant screenshots or GIFs that demonstrate the changes made.

## Additional Notes

Please provide any additional information or context that may be helpful for reviewers.

### CODE REVIEW

File: pallets/subtensor/src/weights.rs [hunk 1]
```diff
@@ -324,4 +378,49 @@ impl<T: Config> Pallet<T> {
...
+
+    pub fn can_commit(netuid: u16, who: &T::AccountId) -> bool {
+        if let Some((_hash, commit_block)) = WeightCommits::<T>::get(netuid, who) {
+            let interval: u64 = Self::get_weight_commit_interval();
+            let current_block: u64 = Self::get_current_block_as_u64();
+            let interval_start: u64 = current_block - (current_block % interval);
+            let last_commit_interval_start: u64 = commit_block - (commit_block % interval);
+
+            // Allow commit if we're within the interval bounds
+            if current_block >= interval_start
```
Comment (gztensor) [inline pallets/subtensor/src/weights.rs]: current_block >= interval_start seems to be always true

File: pallets/subtensor/src/weights.rs [hunk 2]
```diff
@@ -1,8 +1,62 @@
...
+                maybe_commit.take().ok_or(Error::<T>::NoCommitFound)?;
+
+            ensure!(
+                Self::is_reveal_block_range(commit_block),
+                Error::<T>::InvalidRevealTempo
+            );
+
+            let provided_hash: H256 = BlakeTwo256::hash_of(&(
+                who.clone(),
+                nonce,
```
Comment (Rapperswil) [inline pallets/subtensor/src/weights.rs:48]: if you include the account nonce in the hash, is the caller allowed to do any transaction between the commit and the reveal ?

Comment (JohnReedV) [inline pallets/subtensor/src/weights.rs]: We have decided to remove the nonce from the hash. Thank you for contributing.

**Comment by sam0x17** [comment]
@JohnReedV clippy fixes

**Review by sam0x17 — APPROVED** [review]

File: pallets/subtensor/src/lib.rs [hunk 3]
```diff
@@ -1196,6 +1218,32 @@ pub mod pallet {
             Self::do_set_weights(origin, netuid, dests, weights, version_key)
         }
 
+        #[pallet::call_index(96)]
+        #[pallet::weight((Weight::from_parts(10_151_000_000, 0)
```
Comment (distributedstatemachine) [inline pallets/subtensor/src/lib.rs]: Can we benchmark these to arrive at the weights ? Intuition says the commits will be cheaper than the previous set_weights function , and reveals more expensive

File: pallets/subtensor/src/weights.rs [hunk 4]
```diff
@@ -1,8 +1,62 @@
 use super::*;
 use crate::math::*;
+use sp_core::H256;
+use sp_runtime::traits::{BlakeTwo256, Hash};
 use sp_std::vec;
 
 impl<T: Config> Pallet<T> {
+    pub fn do_commit_weights(
```
Comment (distributedstatemachine) [inline pallets/subtensor/src/weights.rs:24]: Can we please add rust docs for these

File: pallets/subtensor/src/weights.rs [hunk 5]
```diff
@@ -1,8 +1,62 @@
...
+
+        WeightCommits::<T>::insert(
+            netuid,
+            &who,
+            (commit_hash, Self::get_current_block_as_u64()),
+        );
+        Ok(())
+    }
+
+    pub fn do_reveal_weights(
```
Comment (distributedstatemachine) [inline pallets/subtensor/src/weights.rs:71]: rust docs please

**Review by distributedstatemachine — CHANGES_REQUESTED** [review]
Looks good overall , but i think we should document all functions and benhmark the new calls

File: pallets/subtensor/src/lib.rs [hunk 6]
```diff
@@ -1230,6 +1247,34 @@ pub mod pallet {
             Self::do_commit_weights(origin, netuid, commit_hash)
         }
 
+        // ---- Used to reveal the weights for a previously commited hash.
```
Comment (distributedstatemachine) [inline pallets/subtensor/src/lib.rs]: Can we please use actual rust doc syntax i.e. `///`  ?

**Review by distributedstatemachine — CHANGES_REQUESTED** [review]

**Review by distributedstatemachine — APPROVED** [review]

**Pull request merged**